src/config.rs

```diff
@@ -115,59 +118,157 @@ impl Display for OverrideReason {
+                        .to_string()
+                }
+                ActiveReason::OverrideDB(ref path) => format!(
+                    "the directory override for '{}' specifies an uninstalled toolchain",
+                    utils::canonicalize_path(path, cfg.notify_handler.as_ref()).display(),
+                ),
+                ActiveReason::ToolchainFile(ref path) => format!(
+                    "the toolchain file at '{}' specifies an uninstalled toolchain",
+                    utils::canonicalize_path(path, cfg.notify_handler.as_ref()).display(),
+                ),
+                ActiveReason::Default => {
+                    "the default toolchain does not describe an installed toolchain".to_string()
+                }
+            };
+
+            Err(anyhow!(reason_err)
+                .context(format!("override toolchain '{name}' is not installed")))
+        }
+        result => Ok(result?),
+    }
+}
+
+// Represents a toolchain override from a +toolchain option, RUSTUP_TOOLCHAIN
+// environment variable, or rust-toolchain.toml file etc. Can include
+// components and targets from a rust-toolchain.toml that should be downloaded
+// and installed.
+#[derive(Debug)]
+enum OverrideCfg {
+    PathBased(PathBasedToolchainName),
+    Custom(CustomToolchainName),
+    Official {
+        toolchain: ToolchainDesc,
+        components: Vec<String>,
+        targets: Vec<String>,
+        profile: Option<dist::Profile>,
+    },
 }
 
 impl OverrideCfg {
     fn from_file(cfg: &Cfg, file: OverrideFile) -> Result<Self> {
-        Ok(Self {
-            toolchain: match (file.toolchain.channel, file.toolchain.path) {
-                (Some(name), None) => Some(
-                    (&ResolvableToolchainName::try_from(name)?
-                        .resolve(&cfg.get_default_host_triple()?)?)
-                        .into(),
-                ),
-                (None, Some(path)) => {
-                    if file.toolchain.targets.is_some()
-                        || file.toolchain.components.is_some()
-                        || file.toolchain.profile.is_some()
-                    {
-                        bail!(
-                            "toolchain options are ignored for path toolchain ({})",
-                            path.display()
-                        )
-                    }
-                    // We -do not- support relative paths, they permit trivial
-                    // completely arbitrary code execution in a directory.
-                    // Longer term we'll not support path based toolchains at
-                    // all, because they also permit arbitrary code execution,
-                    // though with more challenges to exploit.
-                    Some((&PathBasedToolchainName::try_from(&path as &Path)?).into())
-                }
-                (Some(channel), Some(path)) => {
+        let toolchain_name: ToolchainName = match (file.toolchain.channel, file.toolchain.path) {
```

Comment:
Here, I'd like a separate commit for the refactoring that splits the `match` for `toolchain_name` out of the `Self` initializer.

Reply:
I second this.